### PR TITLE
Use the new JSX transform in React environment.

### PR DESCRIPTION
**Pull request by Mikilll94**
## Proposed Changes

New JSX transform was introduced in React about 8 months ago: https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html

Currently Create React App and Next.js supports it out of the box.

Thanks to the new JSX transform you don't have to write:
```
import React from "react"
```
at the top of every JSX/TSX file.

After merging this PR, React environment in Bit will use the new JSX transform for compatible React version.

The code in this PR is heavily inspired by ejected Create React App project.

### CODE REVIEW

**Review by its4zahoor — COMMENTED** [review]
It will be better to move `hasJsxRuntime` into some src/util file for reusability.

File: scopes/react/react/jest/transformer.js [hunk 1]
```diff
@@ -1,8 +1,26 @@
 /* eslint-disable global-require */
 const { transform } = require('@babel/core');
 
+const hasJsxRuntime = (() => {
```
Comment (GiladShoham) [inline scopes/react/react/jest/transformer.js:4]: This exact function appears in this PR multiple times.
In this case, you should extract it to its own component and use it from there.
Also, I'm not sure this function is necessary at all. 
I Will explaining further in the below comments

File: scopes/react/react/jest/transformer.js [hunk 2]
```diff
@@ -1,8 +1,26 @@
...
+  if (process.env.DISABLE_NEW_JSX_TRANSFORM === 'true') {
+    return false;
+  }
+
+  try {
+    require.resolve('react/jsx-runtime');
```
Comment (GiladShoham) [inline scopes/react/react/jest/transformer.js:10]: The `react/jsx-runtime` will always be resolved from the react env context, so no necessary to check it.

File: scopes/react/react/jest/transformer.js [hunk 3]
```diff
@@ -1,8 +1,26 @@
...
+    return false;
+  }
+})();
+
 const presets = [
-  require('@babel/preset-react'),
+  [
+    require('@babel/preset-react'),
+    {
+      runtime: hasJsxRuntime ? 'automatic' : 'classic',
```
Comment (GiladShoham) [inline scopes/react/react/jest/transformer.js:21]: as continue to previous comments, since it will always be resolved, you can just use automatic here

File: scopes/react/react/webpack/webpack.config.preview.dev.ts [hunk 4]
```diff
@@ -35,6 +35,19 @@ const moduleFileExtensions = [
   'md',
 ];
 
+const hasJsxRuntime = (() => {
```
Comment (GiladShoham) [inline scopes/react/react/webpack/webpack.config.preview.dev.ts:38]: not necessary, see comment above

File: scopes/react/react/webpack/webpack.config.preview.dev.ts [hunk 5]
```diff
@@ -103,7 +116,12 @@ export default function ({ envId, fileMapPath, workDir }: Options): WebpackConfi
             configFile: false,
             presets: [
               // Preset includes JSX, TypeScript, and some ESnext features
-              require.resolve('babel-preset-react-app'),
+              [
+                require.resolve('babel-preset-react-app'),
+                {
+                  runtime: hasJsxRuntime ? 'automatic' : 'classic',
```
Comment (GiladShoham) [inline scopes/react/react/webpack/webpack.config.preview.dev.ts:122]: use automatic, see above

File: scopes/react/react/webpack/webpack.config.preview.ts [hunk 6]
```diff
@@ -28,6 +28,19 @@ const moduleFileExtensions = [
   'md',
 ];
 
+const hasJsxRuntime = (() => {
```
Comment (GiladShoham) [inline scopes/react/react/webpack/webpack.config.preview.ts:31]: see comments above

File: scopes/react/react/webpack/webpack.config.preview.ts [hunk 7]
```diff
@@ -256,7 +269,14 @@ export default function (fileMapPath: string): Configuration {
                 babelrc: false,
                 configFile: false,
                 customize: require.resolve('babel-preset-react-app/webpack-overrides'),
-                presets: [require.resolve('@babel/preset-react')],
+                presets: [
+                  [
+                    require.resolve('@babel/preset-react'),
+                    {
+                      runtime: hasJsxRuntime ? 'automatic' : 'classic',
```
Comment (GiladShoham) [inline scopes/react/react/webpack/webpack.config.preview.ts:276]: see comments above

**Review by GiladShoham — CHANGES_REQUESTED** [review]

File: scopes/react/react/jest/transformer.js (hunk 1 above)
Comment (GiladShoham) [reply scopes/react/react/jest/transformer.js:4]: this function might be only relevant for the `process.env.DISABLE_NEW_JSX_TRANSFORM` in this case extract it to a central location

Comment (GiladShoham) [reply scopes/react/react/jest/transformer.js:4]: in case you change this funciton to only check the env variable, you should change its name, as it's not hasJsxRuntime any more, more like `useNewJsxTransform`

**Review by GiladShoham — CHANGES_REQUESTED** [review]

**Review by GiladShoham — APPROVED** [review]

**Comment by KutnerUri** [comment]
wow @Mikilll94, thank you for doing this!
This could break some users though, maybe allow users to opt out by configuration, via `react.env.ts`?

I'm thinking - either
1. pass it from `react.main.runtime.ts` (using `ReactMainConfig`), but that would be global to all react components in the workspace
2. add a property `newJsx` (`reactlessJsxTransform`? `nextJsxTransform`?) on `react.env` and allow composing envs to override this value.

We'd also have to support it in `bit.dev` by upgrading react to `v15.7.0`, and enabling it in the `.babelrc` (should be pretty easy)

**Comment by GiladShoham** [comment]
@Mikilll94 
@KutnerUri Is right about affecting existing users who consume their components from react16 project.
We might need to create a new env for react16 only and make the react env, react17 only.

**Comment by KutnerUri** [comment]
TL;DR for the new JSX transform (aka `jsx v3`).
It's supported by React `15.7.0<`, `16.14.0<` and `17`

original:
```tsx
import React from 'react';

function App() {
  return <h1>Hello World</h1>;
}
```
transformed:
```tsx
import React from 'react';

function App() {
  return React.createElement('h1', null, 'Hello world');
}
```

new transform
```tsx
// Inserted by a compiler (don't import it yourself!)
import {jsx as _jsx} from 'react/jsx-runtime';

function App() {
  return _jsx('h1', { children: 'Hello world' });
}
```

This gives nice performance improvements, and should improve prop passing,

**Comment by GiladShoham** [comment]
@KutnerUri If it's supported since version 15.7.0 we should be good to go isn't it?

**Comment by Mikilll94** [comment]
> @KutnerUri If it's supported since version 15.7.0 we should be good to go isn't it?

@GiladShoham
This is not exactly true. It's supported by React 15.7.0<, 16.14.0< and 17

So i.e. 15.7.0 supports new JSX transform but 16.0.0 do not support it.

**Comment by KutnerUri** [comment]
yes. That's why we need to allow opting-out.
@GiladShoham suggested making a `react16 env`, though it will harder to maintain.

just to make sure the `typescript/tsconfig(.build).json` is making the transform in the dists?

**Comment by GiladShoham** [comment]
not relevant anymore.
The new react env - https://bit.cloud/bitdev/react/react-env already supports it
